File: prodplus-api/src/main/java/br/com/prodplus/models/forms/LoteProdutoForm.java
package br.com.prodplus.models.forms;

import java.io.Serializable;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.HashSet;

import javax.validation.constraints.NotNull;
import javax.validation.constraints.PositiveOrZero;

import br.com.prodplus.models.LoteProduto;
import br.com.prodplus.services.LoteProdutoService;
import br.com.prodplus.services.ProdutoService;
import br.com.prodplus.utils.GeradorRastreio;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

/**
 * 
 * @author Marlon Fernando Garcia
 *
 */
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class LoteProdutoForm implements Serializable {

	private static final long serialVersionUID = 7378683833844448034L;

	private Integer id;
	@NotNull(message = "o produto é obrigatório!")
	private Integer produto;
	@NotNull(message = "a data de produção é obrigatória!")
	private LocalDateTime producao;
	@NotNull(message = "o custo total é obrigatório!")
	@PositiveOrZero(message = "o custo total é inválido!")
	private BigDecimal custoTotal;
	@NotNull(message = "a quantidade inicial é obrigatória!")
	@PositiveOrZero(message = "quantidade inicial inválida!")
	private Double quantInicial;
	@NotNull(message = "a quantidade atual é obrigatória!")
	@PositiveOrZero(message = "quantidade atual inválida!")
	private Double quantAtual;
	@NotNull(message = "campo obrigatório!")
	private boolean ativo = true;

	public LoteProdutoForm(LoteProduto lote) {
		this.setId(lote.getId());
		this.setProduto(lote.getProduto().getId());
		this.setProducao(lote.getProducao());
		this.setCustoTotal(lote.getCustoTotal());
		this.setQuantInicial(lote.getQuantInicial());
		this.setQuantAtual(lote.getQuantAtual());
		this.setAtivo(lote.isAtivo());
	}

	public LoteProduto converter(ProdutoService produtoService, LoteProdutoService loteService) {
		if (this.getId() != null) {
			LoteProduto lote = loteService.buscar(this.getId());
			lote.setAtivo(this.isAtivo());
			lote.setCustoTotal(this.getCustoTotal());
			lote.setProducao(this.getProducao());
			lote.setProduto(produtoService.buscar(this.getProduto()));
			lote.setQuantAtual(this.getQuantAtual());
			lote.setQuantInicial(this.getQuantInicial());
			lote.setRastreio(GeradorRastreio.loteProduto(lote));
			return lote;
		} else {
			return new LoteProduto(null, null, produtoService.buscar(this.getProduto()),
					this.getProducao(), this.getCustoTotal(), this.getQuantInicial(),
					this.getQuantAtual(), new HashSet<>(), new HashSet<>(), this.isAtivo());
		}
	}

}

File: prodplus/src/app/models/forms/lote-produto-form.ts
import { toDateApi } from 'src/app/shared/date-utils';
import { LoteProduto } from '../lote-produto';

export class LoteProdutoForm {
  id: number | null;
  produto: number;
  producao: string;
  custoTotal: number;
  quantInicial: number;
  quantAtual: number;
  ativo: boolean;

  constructor();
  constructor(l: LoteProduto);
  constructor(
    produto: number,
    producao: string,
    custoTotal: number,
    quantInicial: number,
    quantAtual: number,
    ativo: boolean,
    id?: number
  );
  constructor(...args: any[]) {
    if (args.length == 0) {
      this.id = null;
      this.produto = 0;
      this.producao = toDateApi(new Date());
      this.custoTotal = 0;
      this.quantInicial = 0;
      this.quantAtual = 0;
      this.ativo = true;
    } else if (args.length == 1) {
      const lote: LoteProduto = args[0];
      this.id = lote.id;
      this.produto = lote.produto.id != null ? lote.produto.id : 0;
      this.producao = lote.producao;
      this.custoTotal = lote.custoTotal;
      this.quantInicial = lote.quantInicial;
      this.quantAtual = lote.quantAtual;
      this.ativo = lote.ativo;
    } else {
      this.produto = args[0];
      this.producao = args[1];
      this.custoTotal = args[2];
      this.quantInicial = args[3];
      this.quantAtual = args[4];
      this.ativo = args[5];
      if (typeof args[6] === 'undefined') this.id = null;
      else this.id = args[6];
    }
  }
}
